lib/charms/observability_libs/v0/kubernetes_compute_resources_patch.py

```diff
@@ -129,145 +134,88 @@ def setUp(self, *unused):
+    requests = sanitize_resource_spec_dict(requests) or {}
+
+    if adhere_to_requests:
+        # Keep limits fixed when `limits` is too low
+        adjusted, fixed = limits.copy(), requests.copy()
+        func = max
+    else:
+        # Pull down requests when limit is too low
+        fixed, adjusted = limits.copy(), requests.copy()
+        func = min
+
+    # adjusted = {}
+    for k in adjusted:
+        if k not in fixed:
+            # The resource constraint is present in the "adjusted" dict but not in the "fixed"
+            # dict. Keep the "adjusted" value as is
+            continue
+
+        adjusted_value = func(parse_quantity(fixed[k]), parse_quantity(adjusted[k]))  # type: ignore[type-var]
+        adjusted[k] = (
+            str(adjusted_value.quantize(decimal.Decimal("0.001"), rounding=decimal.ROUND_UP))  # type: ignore[union-attr]
+            .rstrip("0")
+            .rstrip(".")
+        )
 
-    With this function:
-    - When the "limits" portion is high enough, the "requests" portion is scaled down
-      proportionally, leaving some room for bursts.
-    - When the "limits" portion is too low, no scaling takes place and the requests equals the
-      limits.
+    return (
+        ResourceRequirements(limits=adjusted, requests=fixed)
+        if adhere_to_requests
+        else ResourceRequirements(limits=fixed, requests=adjusted)
+    )
 
-    Args:
-        resource_limits: A dictionary representation of K8s resource limits
-        default_requests: The default requests values to use if a limits value is not given. This
-            would also be used as the crossover point from requests = limits to the scaled limits.
-            For example, we can set the "requests" portion of the resource limits to a sensible
-            value, while keeping the "limits" portion unspecified.
-        scaling_factor: A scaling factor used to calculate the requests value from the limits
-            value. The scaling factor must satisfy: 0 < scaling_factor < 1.
 
-    Raises:
-        ValueError, if arguments are invalid or out of range.
+def is_valid_spec(spec: Optional[dict], debug=False) -> bool:  # noqa: C901
+    """Check if the spec dict is valid.
 
-    Returns:
-        A resource spec dictionary scaled-down by scaling_factor, subject to the "default value"
-         constraint as described above.
-
-    >>> sf = Decimal("0.8")
-    >>> limits_to_requests_scaled({"cpu": "1"}, {"cpu": "200m"}, sf)
-    {'cpu': '0.800'}
-    >>> limits_to_requests_scaled({"cpu": "260m"}, {"cpu": "200m"}, sf)
-    {'cpu': '0.208'}
-    >>> limits_to_requests_scaled({"cpu": "250m"}, {"cpu": "200m"}, sf)
-    {'cpu': '0.200'}
-    >>> limits_to_requests_scaled({"cpu": "200m"}, {"cpu": "200m"}, sf)
-    {'cpu': '0.200'}
-    >>> limits_to_requests_scaled({"cpu": "150m"}, {"cpu": "200m"}, sf)
-    {'cpu': '0.150'}
+    TODO: generally, the keys can be anything, not just cpu and memory. Perhaps user could pass
```

Comment:
I think it's totally fine to limit it to only CPU and memory for now.

Reply:
Lightkube [doesn't restrict](https://github.com/gtsystem/lightkube/pull/37#discussion_r916034989) it so I also opted to not restrict key values. I added [a doctest](https://github.com/canonical/observability-libs/blob/001b3aa6c70704668d5ada96667b2b1bc645a3c7/lib/charms/observability_libs/v0/kubernetes_compute_resources_patch.py#L174) for a custom resource.
If you believe we _should_ restrict, I am happy to change this.